In [ ]:
import pandas as pd
pd.set_option("display.precision", 2)

In [ ]:
!wget -P ./../data/ https://raw.githubusercontent.com/pbonillor/DATA_SCIENCE_CON_PYTHON/main/MODULO_I/titanic_train.csv

--2023-11-10 02:23:18--  https://raw.githubusercontent.com/pbonillor/DATA_SCIENCE_CON_PYTHON/main/MODULO_I/titanic_train.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 60302 (59K) [text/plain]
Saving to: ‘./../data/titanic_train.csv’

titanic_train.csv   100%[===================>]  58.89K  --.-KB/s    in 0.04s   

2023-11-10 02:23:19 (1.29 MB/s) - ‘./../data/titanic_train.csv’ saved [60302/60302]



In [ ]:
data = pd.read_csv('../data/titanic_train.csv', index_col='PassengerId')

In [ ]:
data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket   Fare Cabin Embarked  
PassengerId                                                        
1                1      0         A/5 21171   7.25   NaN        S  
2                1      0          PC 17599  71.28   C85        C  
3                0      0  STON/O2. 3101282   7.92   NaN        S  
4                1      0            113803  53.10  C123        S  
5                0      0            373450   8.05   NaN        S

In [ ]:
data.describe()

Survived  Pclass     Age   SibSp   Parch    Fare
count    891.00  891.00  714.00  891.00  891.00  891.00
mean       0.38    2.31   29.70    0.52    0.38   32.20
std        0.49    0.84   14.53    1.10    0.81   49.69
min        0.00    1.00    0.42    0.00    0.00    0.00
25%        0.00    2.00   20.12    0.00    0.00    7.91
50%        0.00    3.00   28.00    0.00    0.00   14.45
75%        1.00    3.00   38.00    1.00    0.00   31.00
max        1.00    3.00   80.00    8.00    6.00  512.33

In [ ]:
data.describe(include=['object', 'bool'])

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

In [ ]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
259                                         Ward, Miss. Anna  female  35.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
378                                Widener, Mr. Harry Elkins    male  27.0   

             SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                            
119              0      1  PC 17558  247.52          B58 B60        C  
259              0      0  PC 17755  512.33              NaN        C  
300              0      1  PC 17558  247.52          B58 B60        C  
312              2      2  PC 17608  262.38  B57 B59 B63 B66        C  
378              0      2    113503  211.50              C82        C

In [ ]:
data[(data['Embarked'] == 'C') &
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                                  
259          35.0      0      0  PC 17755  512.33              NaN        C  
680          36.0      0      1  PC 17755  512.33      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.33             B101        C  
312          18.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C

In [ ]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [ ]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories

In [ ]:
data['Age_category'] = data['Age'].apply(age_category)

# 1. How many men/women were there onboard?
- 412 men and 479 women
- 314 men and 577 women
- 479 men and 412 women
- **<font color='green'>577 men and 314 women</font>**

In [ ]:
# Option 1
data['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [ ]:
# Option 2
(data['Sex']=='male').sum(), (data['Sex']=='female').sum()

(577, 314)

# 2. How many men from second class were there onboard?

**Tip: Print the distribution of the `Pclass` feature with the `Sex` feature.**
- 104
- **<font color='green'>108</font>**
- 112
- 125

In [ ]:
# Solution with filters and then count rows
len( data[(data.Sex=="male") & (data['Pclass']==2)] )

108

In [ ]:
# Solution crosstab
pd.crosstab(data['Pclass'], data['Sex'], margins=True)

Sex     female  male  All
Pclass                   
1           94   122  216
2           76   108  184
3          144   347  491
All        314   577  891

# 3. What are median and standard deviation of `Fare`?
- **<font color='green'>median is  14.45, standard deviation is 49.69</font>**
- median is 15.1, standard deviation is 12.15
- median is 13.15, standard deviation is 35.3
- median is  17.43, standard deviation is 39.1

In [ ]:
print("Fare median:", round( data['Fare'].median(), 2))
print("Fare std:   ", round( data['Fare'].std(),    2))

Fare median: 14.45
Fare std:    49.69


# 4. Is the mean age of survived people higher than people who died?
- Yes, the mean age of survived people is higher.
- **<font color='green'>No, the mean age of survived people is lower.</font>**

In [ ]:
# Solution simple
print("Died people, mean age:    ", data[data['Survived'] == 0].Age.mean() )
print("Survived people, mean age:", data[data['Survived'] == 1].Age.mean() )

Died people, mean age:     30.62617924528302
Survived people, mean age: 28.343689655172415


In [ ]:
# Solution with group by
data.groupby('Survived')['Age'].mean()

Survived
0    30.63
1    28.34
Name: Age, dtype: float64

# 5. What are the percentages of survived passengers for:
## a) younger than 30 y.o.
## b) older than 60 y.o.

- 22.7% among young and 40.6% among old
- **<font color='green'>40.6% among young and 22.7% among old</font>**
- 35.3% among young and 27.4% among old
- 27.4% among young and  35.3% among old

In [ ]:
young = data[data['Age'] < 30]
old   = data[data['Age'] > 60]

print("Survived young people:", round(100 * young.Survived.mean(), 1), "%")
print("Survived old people:  ", round(100 * old.Survived.mean(), 1), "%")

Survived young people: 40.6 %
Survived old people:   22.7 %


# 6. Is that true that women survived more frequently than men? What are the percentages?
- 30.2% among men and 46.2% among women
- 35.7% among men and 74.2% among women
- 21.1% among men and 46.2% among women
- **<font color='green'>18.9% among men and 74.2% among women</font>**

In [ ]:
# Solution simple
women = data[data['Sex']=="female"]
men   = data[data['Sex']=="male"]

print("Survived women:", round(100 * women.Survived.mean(), 1), "%")
print("Survived men:  ", round(100 * men.Survived.mean(), 1), "%")

Survived women: 74.2 %
Survived men:   18.9 %


In [ ]:
# Solution with group by
data.groupby('Sex')['Survived'].mean()

Sex
female    0.74
male      0.19
Name: Survived, dtype: float64

# 7. What's the most popular first name among male passengers?
- Charles
- Thomas
- **<font color='green'>William</font>**
- John

In [ ]:
def getNameFunction(fullname):
    return fullname.split(",")[1].split(" ")[2]

In [ ]:
men = data[data['Sex'] == 'male']
men["Name"].apply(getNameFunction).value_counts().head()

William    35
John       25
George     14
Charles    13
Thomas     13
Name: Name, dtype: int64

# 8. How is average age for men/women dependent on `Pclass`?
### Choose all correct statements:
- **<font color='green'>On average, men of 1 class are older than 40</font>**
- On average, women of 1 class are older than 40
- **<font color='green'>Men of all classes are on average older than women of the same class</font>**
- **<font color='green'> On average, passengers ofthe first class are older than those of the 2nd class who are older than passengers of the 3rd class</font>**

In [ ]:
data.pivot_table(['Age'], ['Pclass', "Sex"], aggfunc='mean').head()

Age
Pclass Sex          
1      female  34.61
       male    41.28
2      female  28.72
       male    30.74
3      female  21.75

In [ ]:
pd.crosstab(data['Pclass'], data['Sex'], values=data['Age'], aggfunc='mean')

Sex     female   male
Pclass               
1        34.61  41.28
2        28.72  30.74
3        21.75  26.51